lib/managementClient.ts

```diff
@@ -0,0 +1,231 @@
+    if (!partitionId || (partitionId && typeof partitionId !== "string" && typeof partitionId !== "number")) {
+      throw new Error("'partitionId' is a required parameter and must be of type: 'string' | 'number'.");
+    }
+    const info: any = await this._makeManagementRequest(connection, Constants.partition, partitionId);
+    const partitionInfo: EventHubPartitionRuntimeInformation = {
+      beginningSequenceNumber: info.begin_sequence_number,
+      hubPath: info.name,
+      lastEnqueuedOffset: info.last_enqueued_offset,
+      lastEnqueuedTimeUtc: new Date(info.last_enqueued_time_utc),
+      lastSequenceNumber: info.last_enqueued_sequence_number,
+      partitionId: info.partition,
+      type: info.type
+    };
+    debug(`[%s] The partition info is: ${partitionInfo}.`, connection.options.id);
+    return partitionInfo;
+  }
+
+  /**
+   * Closes the AMQP management session to the Event Hub for this client,
+   * returning a promise that will be resolved when disconnection is completed.
+   * @return {Promise<void>}
+   */
+  async close(): Promise<void> {
+    try {
+      if (this._mgmgtReqResLink) {
+        await rheaPromise.closeSession(this._mgmgtReqResLink.session);
+        debug("Successfully closed the management session.");
+        this._mgmgtReqResLink = undefined;
+      }
+    } catch (err) {
+      const msg = `An error occurred while closing the management session: ${err}`;
+      debug(msg);
+      return Promise.reject(msg);
+    }
+  }
+
+  private async _init(connection: any, endpoint: string, replyTo: string): Promise<void> {
+    if (!this._mgmgtReqResLink) {
+      const rxopt: rheaPromise.ReceiverOptions = { source: { address: endpoint }, name: replyTo, target: { address: replyTo } };
+      debug("Creating a session for $management endpoint");
+      this._mgmgtReqResLink = await createRequestResponseLink(connection, { target: { address: endpoint } }, rxopt);
+      debug(`[${connection.options.id}] Created sender "${this._mgmgtReqResLink.sender.name}" and receiver "${this._mgmgtReqResLink.receiver.name}" links for $management endpoint.`);
+    }
+  }
+
+  /**
+   * @private
+   * Helper method to make the management request
+   * @param {Connection} connection - The established amqp connection
+   * @param {string} type - The type of entity requested for. Valid values are "eventhub", "partition"
+   * @param {string | number} [partitionId] - The partitionId. Required only when type is "partition".
+   */
+  private async _makeManagementRequest(connection: any, type: "eventhub" | "partition", partitionId?: string | number): Promise<any> {
+    if (partitionId && typeof partitionId !== "string" && typeof partitionId !== "number") {
+      throw new Error("'partitionId' is a required parameter and must be of type: 'string' | 'number'.");
+    }
+    return new Promise(async (resolve: any, reject: any) => {
```

Comment:
i typically find that it's better to wrap the smallest possible amount of code in `new Promise()`, especially when you're using async/await. that way you don't have to worry about try/catching your `async` calls inside the promise callback and the extra level of indirection caused by `new Promise()` is as localized as possible.

in this case, you should just be able to wrap the `messageCallback` and the sender send/receiver listener bits in the promise constructor and remove the `async` from the internal callback

Reply:
cool 👍